itest/litd_custom_channels_test.go

```diff
@@ -2046,6 +2050,166 @@ func testCustomChannelsLiquidityEdgeCases(_ context.Context,
 
 	logBalance(t.t, nodes, assetID, "after failed payment (asset "+
 		"invoice, strict forwarding)")
+
+	// Edge case: Fabia gets all the asset liquidity on her side. Then
+	// generates an asset invoice to be paid for with assets by Charlie.
+	// Charlie will unleash multiple shards towards Fabia, exhausting the
+	// liquidity in the Erin<->Fabia channel. Dave should be able to detect
+	// those failures and account for those cancelled HTLCs in the RFQ
+	// tracking mechanism. Eventually we will slosh some funds from Fabia
+	// back to Erin in order for Erin to be able to forward the rest of the
+	// HTLCs, leading to the release of the preimage.
+
+	// Erin starts by sending half of the assets to Fabia's side. He also
+	// sends some raw sats to be used by Fabia later for the sloshing.
+	sendAssetKeySendPayment(
+		t.t, erin, fabia, 150_000, assetID,
+		fn.None[int64](), lnrpc.Payment_SUCCEEDED,
+		fn.None[lnrpc.PaymentFailureReason](),
+	)
+	sendKeySendPayment(t.t, erin, fabia, 20_000)
+
+	logBalance(t.t, nodes, assetID, "balance after 1st slosh")
+
+	// Fabia creates an asset invoice of 100k assets. There is currently
+	// not enough asset liquidity on Erin's side to forward HTLCs to satisfy
+	// this invoice.
+	invoiceResp = createAssetInvoice(t.t, erin, fabia, 100_000, assetID)
+
+	// We set the waiting period for the slosh payment to occur. This is set
+	// to half of the default payment timeout, as we want it to occur half
+	// way through the in-flight payment.
+	sloshWait := PaymentTimeout / 3
+
+	go func() {
+		// After a small delay (less than the payment timeout) Fabia
+		// sloshes back the asset liquidity to Erin. This should allow
+		// the payment by Charlie to eventually complete.
+		time.Sleep(sloshWait)
+		sendAssetKeySendPayment(
+			t.t, fabia, erin, 75_000, assetID,
+			fn.None[int64](), lnrpc.Payment_SUCCEEDED,
+			fn.None[lnrpc.PaymentFailureReason](),
+		)
+	}()
+
+	// To avoid goroutine uncertainty, we wait as much as the above routine
+	// minus a small delta. This is enough for us to be sure that the
+	// payment will complete due to the slosh.
```

Comment:
This is a bit tricky. Under the hood we know that some HTLCs of some size are going to be forwarded and eventually cancelled. If we were to intercept (which adds some overhead in code complexity) we would also run into the issue of not knowing what to do with the HTLC we just intercepted. The reason the time based approach is nice here is because it's agnostic to the internal mechanisms of sender/router/receiver.

We don't know what's going on inside, but we want:
a) Sender to try to send payment and have their attempts failing
b) Slosh payment to complete ~halfway through the payment timeout
c) Sender to eventually complete payment

We could intercept HTLCs to guarantee that any attempt was made during a), but other than that could we use it any different way?

Reply:
You call into `payInvoiceWithAssets` below this point. And internally that handles a stream:
https://github.com/lightninglabs/lightning-terminal/blob/e52a8c563321f4fcc71c0e5bb22da56f0531a6dc/itest/assets_test.go#L888

Could you perhaps call tapd `SendPayment` in your test and then monitor the stream?

I think that the stream should have a payment result payment status of in-flight before sloshing. And then after sloshing you can listen on the stream for a payment result payment status success.

That way you can be sure that the send attempt is in-flight and then is successful after sloshing. And without timing complexity or perhaps any extra goroutines?

Reply:
After latest foce-push I'm using a new approach, so PTAL